packages/runtime/container-runtime/src/summaryGenerator.ts

```diff
@@ -276,6 +277,26 @@ export class SummaryGenerator {
                 return fail("submitSummaryFailure", summaryData.error, generateTelemetryProps);
             }
 
+            /**
+             * With incremental summaries, if the full tree was not summarized, only data stores that changed should
+             * be summarized. A data store is considered changed if either or both of the following is true:
+             * - It has received an op.
+             * - Its reference state changed, i.e., it went from referenced to unreferenced or vice-versa.
+             *
+             * Since an op in a data store can change the reference state of another data store, each op can potentially
+             * result in two data stores being summarized. So, the number of summarized data stores should not exceed
+             * twice the number of ops since last summary.
+             */
+            if (!fullTree
+                && !summaryData.forcedFullTree
+                && summaryData.summaryStats.summarizedDataStoreCount <= 2 * opsSinceLastSummary) {
+                logger.sendErrorEvent({
```

Comment:
@vladsud - I logged an error here instead of an assert for 2 reasons:
1. It's possible that the bug fixed by this PR isn't the only one resulting in large number of summarized data stores. So, we should monitor telemetry before converting this to an assert.
2. Theoretically, multiple handles can be added to a DDS via an object. This would result in one op changing multiple data stores. Although, practically this check shouldn't get violated as we have of buffer. But being safe for now.

Reply:
Actually, there a 3rd reason why we cannot do this as pointed out in the issue https://github.com/microsoft/FluidFramework/issues/8963#issuecomment-1029427346.